subprojects/core-utils/src/main/java/dev/nokee/utils/NamedDomainObjectCollectionUtils.java

```diff
@@ -72,24 +90,65 @@ private NamedDomainObjectCollectionUtils() {}
 	//endregion
 
 	//region whenElementKnown
-	public static <T> void whenElementKnown(NamedDomainObjectCollection<T> self, Action<? super ElementInfo<T>> action) {
-		((DefaultNamedDomainObjectCollection<T>) self).whenElementKnown(compose(action, ElementInfo::of));
+	/**
+	 * Adds callback when new element is known to the named collection.
+	 * This API uses the {@link DefaultNamedDomainObjectCollection#whenElementKnown(Action)} internal API.
+	 * The Gradle API can call back twice, once when registering the element and another time when the element is created.
+	 * This API deduplicate callback execution to ensure a single callback per element.
+	 *
+	 * @param self  the collection, must not be null
+	 * @param action  the action to call for each known element, must not be null
+	 * @param <T>  the base element type
+	 */
+	public static <T> void whenElementKnown(NamedDomainObjectCollection<T> self, Action<? super KnownElement<T>> action) {
+		((DefaultNamedDomainObjectCollection<T>) self).whenElementKnown(new ToKnownElementAction<T>(action) {
+			@Override
+			protected KnownElement<T> create(DefaultNamedDomainObjectCollection.ElementInfo<T> elementInfo) {
+				return KnownElement.of(self, elementInfo);
+			}
+		});
+	}
+
+	private static abstract class ToKnownElementAction<T> implements Action<DefaultNamedDomainObjectCollection.ElementInfo<T>> {
+		private final Set<String> deduplicateElements = new HashSet<>();
+		private final Action<? super KnownElement<T>> action;
+
+		ToKnownElementAction(Action<? super KnownElement<T>> action) {
+			this.action = action;
+		}
+
+		@Override
+		public final void execute(DefaultNamedDomainObjectCollection.ElementInfo<T> elementInfo) {
+			if (deduplicateElements.add(elementInfo.getName())) {
+				action.execute(create(elementInfo));
+			}
+		}
+
+		protected abstract KnownElement<T> create(DefaultNamedDomainObjectCollection.ElementInfo<T> elementInfo);
+	}
+	//endregion
+
+	//region getType
+	public static <T> Class<?> getType(NamedDomainObjectCollection<T> self) {
+		return ((DefaultNamedDomainObjectCollection<T>) self).getType();
```

Comment:
❌ I don't think this API is named properly. It followed the internal API name but even that I'm unsure if it's correct. The intention here is to get the base element type of the container, i.e. `getBaseElementType()` or `getElementBaseType()`. Regardless of what we choose here, `getType()` is too vague.